tokenomics_decentralization/helper.py

```diff
@@ -468,3 +443,161 @@ def get_usd_cent_equivalent(ledger, date):
+        csv_writer.writerows(output_rows)
+
+
+def get_active_source_keywords():
+    """
+    Returns the keywords of the sources that should be used in the analysis based on the config parameters.
+    :returns: a list of strings of mapping information source keywords
+    """
+    try:
+        return get_config_data()['analyze_flags']['clustering_sources']
+    except KeyError:
+        raise ValueError('Clustering sources does not exist in analyze flags')
+
+
+def get_active_sources():
+    """
+    Returns the sources that should be used in the analysis based on the config parameters.
+    :returns: a list of strings of mapping information sources
+    """
+    with open(MAPPING_INFO_DIR / 'sources.json') as f:
+        source_keywords = json.load(f)
+
+    active_sources = set()
+    for kw in get_active_source_keywords():
+        for source in source_keywords[kw]:
+            active_sources.add(source)
+
+    return active_sources
+
+
+def get_clusters(ledger):
+    """
+    Retrieves the clusters of addresses that form from the mapping information.
+    First identifies the addresses that are associated (in the mapping
+    information) with more than one entities.
+    Then it retrieves the set of all entities that each such address is
+    associated with and constructs the clusters by finding overlapping entities in these sets.
+    Finally it constructs a dictionary that maps entities to their cluster.
+    :param ledger: a string of the ledger's name
+    :returns: a dictionary where the key is an entity and the value is the cluster to which it belongs
+    """
+    # Find addresses that are associated with more than one entities.
+    # Note: If the mapping file is very large, loading the whole dictionary can result to running out of memory. Using this step consumes less memory but requires one extra pass of the file.
+    addresses = set()
+    multi_entity_addresses = set()
+    with open(MAPPING_INFO_DIR / f'addresses/{ledger}.jsonl') as f:
+        for line in f:
+            info = json.loads(line)
+            address = info['address']
+            if address in addresses:
+                multi_entity_addresses.add(address)
+            addresses.add(address)
+    del addresses
+
+    # Get the entities associated with the multi-entity addresses.
+    address_entities = defaultdict(set)
+    with open(MAPPING_INFO_DIR / f'addresses/{ledger}.jsonl') as f:
+        for line in f:
+            info = json.loads(line)
+            address = info['address']
+            if address in multi_entity_addresses:
+                source = info['source']
+                entity_name = info['name']
+                if source not in get_active_sources():
+                    continue
+                address_entities[address].add((entity_name, source))
+    del multi_entity_addresses
+
+    clusters = list(address_entities.values())
+    del address_entities
+
+    # If an entity is present in two entries, then these are merged.

```

Comment:
I don't entirely understand this process. In general, the way we merge information from different sources seems like something that should be more documented, so perhaps worth creating a new page in our docs about the mapping process, where this could also be explained in more detail (and with examples) to make it easier to understand

Reply:
This also seems like a fragment of code that could be in a separate function, as the logic here is about merging clusters and not just retrieving them as the current function it belongs in suggests